File: Language/C#/Book/tmp/LINQ实战/Chapter05.CSharp/Chapter05/Extensions.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace LinqInAction.Chapter05
{
    public static class Extensions
    {
        public static void ForEach<T>(this IEnumerable<T> source, Action<T> func)
        {
            foreach (var item in source)
                func(item);
        }

        public static TElement MaxElement<TElement, TData>(this IEnumerable<TElement> source, Func<TElement, TData> selector) where TData : IComparable<TData>
        {
            if (source == null)
                throw new ArgumentNullException("source");
            if (selector == null)
                throw new ArgumentNullException("selector");

            Boolean firstElement = true;
            TElement result = default(TElement);
            TData maxValue = default(TData);
            foreach (TElement element in source)
            {
                var candidate = selector(element);
                if (firstElement || (candidate.CompareTo(maxValue) > 0))
                {
                    firstElement = false;
                    maxValue = candidate;
                    result = element;
                }
            }
            return result;
        }
    }
}

File: Language/C#/Book/tmp/LINQ实战/Chapter05.VB/Chapter05/Extensions.vb
Imports System.Runtime.CompilerServices

Module Extensions
  <Extension()> _
    Public Sub ForEach(Of T)(ByVal source As IEnumerable(Of T), ByVal func As Action(Of T))
    For Each item In source
      func(item)
    Next
  End Sub

  <Extension()> _
    Public Function MaxElement(Of TElement, TData As IComparable(Of TData))( _
      ByVal source As IEnumerable(Of TElement), _
      ByVal selector As Func(Of TElement, TData)) As TElement
    If source Is Nothing Then
      Throw New ArgumentNullException("source")
    End If
    If selector Is Nothing Then
      Throw New ArgumentNullException("selector")
    End If

    Dim firstElement As Boolean = True
    Dim result As TElement ' = default(TElement)
    Dim maxValue As TData ' = default(TData)
    For Each element In source
      Dim candidate = selector(element)
      If firstElement Or (candidate.CompareTo(maxValue) > 0) Then
        firstElement = False
        maxValue = candidate
        result = element
      End If
    Next
    Return result
  End Function
End Module